src/test/java/org/apache/sysml/test/gpu/NeuralNetworkOpTests.java

```diff
@@ -0,0 +1,490 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.sysml.test.gpu;
+
+import org.apache.sysml.api.mlcontext.Matrix;
+import org.apache.sysml.runtime.util.ConvolutionUtils;
+import org.apache.sysml.test.utils.TestUtils;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * Test neural network operations on the GPU
+ */
+public class NeuralNetworkOpTests extends GPUTests {
+
+	private final static String TEST_NAME = "NeuralNetworkOpTests";
+	private final int seed = 42;
+
+	// The MAX_OP_SIZE is to take into consideration the memory available on the GPU as well as
+	// limits set by cudnn (operands need to be less than 2GB)
+	private final double MAX_OP_SIZE = 1.5 * 1024 * 1024 * 1024; // 1 GB
+
+	// More comprehensive but time consuming tests
+	/*
+	private final List<Integer> Nlst = Arrays.asList(128, 64, 32);
+    private final List<Integer> Clst = Arrays.asList(30, 20, 3);
+    private final List<Integer> Hlst = Arrays.asList(400, 128, 32);
+    private final List<Integer> Wlst = Arrays.asList(400, 128, 32);
+    private final List<Integer> Klst = Arrays.asList(30, 20, 10);
+    private final List<Integer> Rlst = Arrays.asList(128, 63, 4);
+    private final List<Integer> Slst = Arrays.asList(128, 63, 4);
+    private final List<Integer> strideHeightLst = Arrays.asList(9, 3);
+    private final List<Integer> strideWidthLst = Arrays.asList(9, 3);
+    private final List<Integer> padHeightLst = Arrays.asList(3, 1);
+    private final List<Integer> padWidthLst = Arrays.asList(3, 1);
+    private final List<Double> sparsitylst = Arrays.asList(1.0);    // Only test for dense
+    */
+
+	private final List<Integer> Nlst = Arrays.asList(128, 64);
+	private final List<Integer> Clst = Arrays.asList(30, 3);
+	private final List<Integer> Hlst = Arrays.asList(256, 64);
+	private final List<Integer> Wlst = Arrays.asList(256, 64);
+	private final List<Integer> Klst = Arrays.asList(30, 20);
+	private final List<Integer> Rlst = Arrays.asList(128, 4);
+	private final List<Integer> Slst = Arrays.asList(128, 4);
```

Comment:
Can you update the sizes lists for `R` and `S` to include `3`?  Maybe just replace 4 with 3.  3x3 filters are super common, so it would be good to have explicit tests for that size.  Also, 1x1 would be good too, since that is used to alter the number of channels only, and I could imagine the engine having an issue for that size.  128 is good too since it's possible to formulate an affine layer (i.e. "fully-connected layer", "dense layer", etc.) as a convolution with filter spatial sizes equal to the input spatial sizes.